File: flutter/craftd_module/lib/presentation/builder/craftd_builder_manager.dart
import '../ui/button/craftd_button_builder.dart';
import '../ui/empty/craftd_empty_builder.dart';
import '../ui/text/craftd_text_builder.dart';
import 'craftd_builder.dart';

class CraftDBuilderManager {
  static final Map<String, CraftDBuilder> _builders = {
    CraftDButtonBuilder.key: CraftDButtonBuilder(),
    CraftDTextBuilder.key: CraftDTextBuilder(),
  };

  static CraftDBuilder getBuilder(String key,
      {Map<String, CraftDBuilder>? customMap}) {
    if (customMap != null) {
      _builders.addAll(customMap);
    }
    return _builders[key] != null ? _builders[key]! : CraftDEmptyBuilder();
  }
}

File: android_kmp/craftd-compose/src/main/java/com/github/codandotv/craftd/compose/builder/CraftDBuilderManager.kt
package com.github.codandotv.craftd.compose.builder

import androidx.compose.runtime.Immutable
import androidx.compose.runtime.Stable
import com.github.codandotv.craftd.androidcore.presentation.CraftDComponentKey
import com.github.codandotv.craftd.compose.ui.button.CraftDButtonBuilder
import com.github.codandotv.craftd.compose.ui.checkbox.CraftDCheckBoxBuilder
import com.github.codandotv.craftd.compose.ui.empty.CraftDEmptyBuilder
import com.github.codandotv.craftd.compose.ui.text.CraftDTextBuilder

@Stable
@Immutable
class CraftDBuilderManager {
    private val mapBuilder = hashMapOf(
        CraftDComponentKey.BUTTON_COMPONENT.key to CraftDButtonBuilder(),
        CraftDComponentKey.TEXT_VIEW_COMPONENT.key to CraftDTextBuilder(),
        CraftDComponentKey.CHECK_BOX_COMPONENT.key to CraftDCheckBoxBuilder(),
    )
    fun add(vararg arrayCraftDBuilder: CraftDBuilder) : CraftDBuilderManager{
        arrayCraftDBuilder.forEach {
            mapBuilder[it.key] = it
        }
        return this
    }

    fun getBuilder(
        key: String,
        customMap: Map<String, CraftDBuilder>? = null
    ): CraftDBuilder {
        customMap?.let {
            mapBuilder.putAll(customMap)
        }
        return mapBuilder[key] ?: CraftDEmptyBuilder();
    }

}